File: HW/HW6/solution.java
import java.util.List;
import java.util.LinkedList;
import java.util.Arrays;

public class solution {
  public List<List<Integer>> threeSum(int[] nums) {
      List<List<Integer>> ret = new LinkedList();
      Arrays.sort(nums);
      for (int i = 0; i < nums.length - 2; i++){
          if (i == 0 || (i > 0 && nums[i-1] != nums[i])){
              int low = i+1;
              int high = nums.length -1;
              int sum = 0-nums[i];
              
              while(low < high){
                  if (nums[low] + nums [high] == sum){
                      ret.add(Arrays.asList(nums[low], nums[high], nums[i]));
                      while(low < high && (nums[low+1] == nums[low])){
                          low++;
                      }
                      while(low < high && (nums[high-1] == nums[high])){
                          high--;
                      }
                      low++;
                      high--;
                  }
                  
                  else if (nums[low] + nums [high] > sum){
                      high--;
                  }
                  else{
                      low++;
                  }
              }
              
          }
      }
      return ret;
      
      
      
    }
    public void print(List<List<Integer>> ret){
        if(ret.size() == 0){
            System.out.println("Empty");
        }
          for (List<Integer> vals : ret){
              for (Integer sol : vals){
                  System.out.print(sol + " ");
              }
              System.out.println();
          }
      }
    
    public static void main(String args[]){
      //Test 1
        solution test1 = new solution();
        int[] nums = new int[]{0,1,1};
        List<List<Integer>> sol1 = test1.threeSum(nums);
        test1.print(sol1);

        int[] nums2 = new int[]{-5,0,5,10,-10,0}
        List<List<Integer>> sol2 = test1.threeSum(nums2);
        test1.print(sol2);
        
    }
}

File: HW/HW6/solution.py
def threeSum(nums):
    ret =[]
    nums.sort()
    for i in range(0,len(nums) - 2,1):
        if(i == 0 or (i > 0 and nums[i] != nums[i-1])):
            sum = 0-nums[i]
            low = 0
            high = len(nums) -1
            while(low < high):
                if(nums[low] + nums[high] == sum):
                    ret.append([nums[i], nums[low], nums[high]])
                    while(low<high and nums[low+1] == nums[low]):
                        low = low + 1
                    while(low<high and nums[high -1] == nums[high]):
                        high = high - 1
                    low = low + 1
                    high = high -1
                
                elif(nums[low] + nums[high] < sum):
                    low = low + 1
                
                else:
                    high = high -1
    return ret
    
# Test Case Provided
nums1 = [0,1,1]
print(threeSum(nums1))

nums2 = [-5,0,5,10,-10,0]
print(threeSum(nums2))
